Automatically set `text_editor/external/exec_flags` based on the editor specified in `exec_path` text box

ivanskorikov opened the pull request:
Whenever `exec_path` is changed, a signal is emitted. On this signal, a method is called that searches the contents of the `exec_path` field for a predefined pattern using a regex. Upon a valid match, a value for `text_editor/external/exec_flags` is selected and this property is modified.

Values for exec_flags are taken from the [Using an external text editor](https://docs.godotengine.org/en/stable/tutorials/editor/external_editor.html) doc

Closes https://github.com/godotengine/godot-proposals/issues/10631 and closes https://github.com/JetBrains/godot-support/issues/226

ivanskorikov commented:
Squashed commits per Godot Git checklist

van800 commented:
@Calinou would you please review or help to find someone to review?

Calinou commented on editor/editor_properties.cpp:
```diff
@@ -462,10 +462,50 @@ void EditorPropertyPath::_set_read_only(bool p_read_only) {
+	Ref<RegEx> regex;
+	regex.instantiate();
+
+	const String editor_pattern = R"([\\/]((?:[jJ]et[Bb]rains\s*)?[rR]ider(?:\s*(EAP|\d{4}\.\d+|\d{4}\.\d+\s*[Dd]ev)?)?|[vV]isual\s*[sS]tudio\s*[cC]ode|[sS]ubl(ime\s*[tT]ext)?|sublime_text|(g)?vim|emacs|atom|geany|kate|code)(?:\.app|\.exe|\.bat|\.sh)?)";
```
There's a way to make the regex case-insensitive, so you don't need to check for both uppercase and lowercase:

- Use the case-insensitive regex flag.
- Make the source text lowercase before matching.

Calinou commented:
@ivanskorikov Bump 🙂 Can you take a look at the review comments above? If you don't have time anymore, let us know and someone can salvage this PR.

ivanskorikov commented:
@Calinou Apologies, got a bit swept up by daily work and business trips. Thanks for the feedback, I've added the suggested changes, namely simplified the regex to handle only lowercase matches and added support for VSCodium

van800 commented:
@Calinou would you please review?

Calinou commented:
> @Calinou would you please review?

This fails to build when rebased on `master` on macOS:

```
editor/editor_properties.cpp:505:7: error: member access into incomplete type 'RegEx'
  505 |         regex->compile(editor_pattern);
      |              ^
./core/io/logger.h:40:7: note: forward declaration of 'RegEx'
   40 | class RegEx;
      |       ^
editor/editor_properties.cpp:506:6: error: use of undeclared identifier 'RegExMatch'
  506 |         Ref<RegExMatch> editor_match = regex->search(new_path.to_lower());
      |             ^
editor/editor_properties.cpp:506:38: error: member access into incomplete type 'RegEx'
  506 |         Ref<RegExMatch> editor_match = regex->search(new_path.to_lower());
      |                                             ^
./core/io/logger.h:40:7: note: forward declaration of 'RegEx'
   40 | class RegEx;
      |       ^
In file included from editor/editor_properties.cpp:31:
In file included from editor/editor_properties.h:34:
In file included from ./editor/editor_inspector.h:34:
In file included from ./editor/add_metadata_dialog.h:34:
In file included from ./editor/gui/editor_validation_panel.h:34:
In file included from ./scene/gui/panel_container.h:34:
In file included from ./scene/gui/container.h:34:
In file included from ./scene/gui/control.h:35:
In file included from ./core/object/gdvirtual.gen.inc:5:
In file included from ./core/object/script_instance.h:34:
./core/object/ref_counted.h:184:29: error: member access into incomplete type 'RegEx'
  184 |                 if (reference && reference->unreference()) {
      |                                           ^
./core/object/ref_counted.h:198:3: note: in instantiation of member function 'Ref<RegEx>::unref' requested here
  198 |                 unref();
      |                 ^
editor/editor_properties.cpp:501:13: note: in instantiation of member function 'Ref<RegEx>::~Ref' requested here
  501 |         Ref<RegEx> regex;
      |                    ^
./core/io/logger.h:40:7: note: forward declaration of 'RegEx'
   40 | class RegEx;
      |       ^
In file included from editor/editor_properties.cpp:31:
In file included from editor/editor_properties.h:34:
In file included from ./editor/editor_inspector.h:34:
In file included from ./editor/add_metadata_dialog.h:34:
In file included from ./editor/gui/editor_validation_panel.h:34:
In file included from ./scene/gui/panel_container.h:34:
In file included from ./scene/gui/container.h:34:
In file included from ./scene/gui/control.h:34:
In file included from ./core/math/transform_2d.h:34:
In file included from ./core/math/math_funcs.h:34:
In file included from ./core/error/error_macros.h:34:
In file included from ./core/object/object_id.h:34:
In file included from ./core/typedefs.h:46:
In file included from /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/c++/v1/utility:254:
In file included from /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/c++/v1/__utility/cmp.h:16:
In file included from /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/c++/v1/limits:819:
In file included from /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/c++/v1/type_traits:510:
/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/c++/v1/__type_traits/is_trivially_destructible.h:26:38: error: incomplete type 'RegEx' used in type trait expression
   26 |     : public integral_constant<bool, __is_trivially_destructible(_Tp)> {};
      |                                      ^
/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/c++/v1/__type_traits/is_trivially_destructible.h:42:53: note: in instantiation of template class 'std::is_trivially_destructible<RegEx>' requested here
   42 | inline constexpr bool is_trivially_destructible_v = is_trivially_destructible<_Tp>::value;
      |                                                     ^
./core/os/memory.h:139:22: note: in instantiation of variable template specialization 'std::is_trivially_destructible_v<RegEx>' requested here
  139 |         if constexpr (!std::is_trivially_destructible_v<T>) {
      |                             ^
./core/object/ref_counted.h:185:4: note: in instantiation of function template specialization 'memdelete<RegEx>' requested here
  185 |                         memdelete(reference);
      |                         ^
./core/object/ref_counted.h:198:3: note: in instantiation of member function 'Ref<RegEx>::unref' requested here
  198 |                 unref();
      |                 ^
editor/editor_properties.cpp:501:13: note: in instantiation of member function 'Ref<RegEx>::~Ref' requested here
  501 |         Ref<RegEx> regex;
      |                    ^
./core/io/logger.h:40:7: note: forward declaration of 'RegEx'
   40 | class RegEx;
      |       ^
In file included from editor/editor_properties.cpp:31:
In file included from editor/editor_properties.h:34:
In file included from ./editor/editor_inspector.h:34:
In file included from ./editor/add_metadata_dialog.h:34:
In file included from ./editor/gui/editor_validation_panel.h:34:
In file included from ./scene/gui/panel_container.h:34:
In file included from ./scene/gui/container.h:34:
In file included from ./scene/gui/control.h:34:
In file included from ./core/math/transform_2d.h:37:
In file included from ./core/templates/vector.h:43:
./core/os/memory.h:140:10: error: member access into incomplete type 'RegEx'
  140 |                 p_class->~T();
      |                        ^
./core/object/ref_counted.h:185:4: note: in instantiation of function template specialization 'memdelete<RegEx>' requested here
  185 |                         memdelete(reference);
      |                         ^
./core/object/ref_counted.h:198:3: note: in instantiation of member function 'Ref<RegEx>::unref' requested here
  198 |                 unref();
      |                 ^
editor/editor_properties.cpp:501:13: note: in instantiation of member function 'Ref<RegEx>::~Ref' requested here
  501 |         Ref<RegEx> regex;
      |                    ^
./core/io/logger.h:40:7: note: forward declaration of 'RegEx'
   40 | class RegEx;
      |       ^
In file included from editor/editor_properties.cpp:31:
In file included from editor/editor_properties.h:34:
In file included from ./editor/editor_inspector.h:34:
In file included from ./editor/add_metadata_dialog.h:34:
In file included from ./editor/gui/editor_validation_panel.h:34:
In file included from ./scene/gui/panel_container.h:34:
In file included from ./scene/gui/container.h:34:
In file included from ./scene/gui/control.h:35:
In file included from ./core/object/gdvirtual.gen.inc:5:
In file included from ./core/object/script_instance.h:34:
./core/object/ref_counted.h:192:14: error: allocation of incomplete type 'RegEx'
  192 |                 ref(memnew(T(p_params...)));
      |                            ^
./core/os/memory.h:125:53: note: expanded from macro 'memnew'
  125 | #define memnew(m_class) _post_initialize(::new ("") m_class)
      |                                                     ^~~~~~~
editor/editor_properties.cpp:502:8: note: in instantiation of function template specialization 'Ref<RegEx>::instantiate<>' requested here
  502 |         regex.instantiate();
      |               ^
./core/io/logger.h:40:7: note: forward declaration of 'RegEx'
   40 | class RegEx;
      |       ^
In file included from editor/editor_properties.cpp:31:
In file included from editor/editor_properties.h:34:
In file included from ./editor/editor_inspector.h:34:
In file included from ./editor/add_metadata_dialog.h:34:
In file included from ./editor/gui/editor_validation_panel.h:34:
In file included from ./scene/gui/panel_container.h:34:
In file included from ./scene/gui/container.h:34:
In file included from ./scene/gui/control.h:35:
In file included from ./core/object/gdvirtual.gen.inc:5:
In file included from ./core/object/script_instance.h:34:
./core/object/ref_counted.h:80:19: error: member access into incomplete type 'RegEx'
   80 |                                 if (!reference->reference()) {
      |                                               ^
./core/object/ref_counted.h:61:3: note: in instantiation of function template specialization 'Ref<RegEx>::ref_pointer<false>' requested here
   61 |                 ref_pointer<false>(p_from.reference);
      |                 ^
./core/object/ref_counted.h:192:3: note: in instantiation of member function 'Ref<RegEx>::ref' requested here
  192 |                 ref(memnew(T(p_params...)));
      |                 ^
editor/editor_properties.cpp:502:8: note: in instantiation of function template specialization 'Ref<RegEx>::instantiate<>' requested here
  502 |         regex.instantiate();
      |               ^
./core/io/logger.h:40:7: note: forward declaration of 'RegEx'
   40 | class RegEx;
      |       ^
8 errors generated.
scons: *** [editor/editor_properties.macos.editor.arm64.o] Error 1
scons: building terminated because of errors.
```

van800 commented:
What do I need to do to take this over from @ivanskorikov?
New pull request?

Calinou commented:
> What do I need to do to take this over from @ivanskorikov? New pull request?

Yes, I suggest opening a new PR.

van800 commented:
We have solved it by adding me as a contributor.
Waiting for the workflow...

van800 commented:
@Calinou would you please review. Sorry for a lot of mentions today, I am desperate to have this merged at least to master.

KoBeWi commented:
This shouldn't be implemented in EditorPropertyPath, but in EditorSettings.

KoBeWi commented on editor/editor_properties.cpp:
```diff
@@ -496,6 +501,42 @@ String EditorPropertyPath::_get_path_text() {
 }
 
+void EditorPropertyPath::_on_exec_path_changed(String new_path) {
+	Ref<RegEx> regex;
+	regex.instantiate();
+
+	const String editor_pattern = R"([\\/]((?:jetbrains\s*)?rider(?:\s*(eap|\d{4}\.\d+|\d{4}\.\d+\s*dev)?)?|visual\s*studio\s*code|subl(ime\s*text)?|sublime_text|(g)?vim|emacs|atom|geany|kate|code|(vs)?codium)(?:\.app|\.exe|\.bat|\.sh)?)";
```
You can use `Regex::create_from_string()`

van800 committed:
Add dynamic exec flags update when exec path is set for an external text editor.
Co-authored-by: Tomasz Chabora <kobewi4e@gmail.com>
Co-authored-by: A Thousand Ships <96648715+AThousandShips@users.noreply.github.com>

van800 commented:
> This shouldn't be implemented in EditorPropertyPath, but in EditorSettings.

moved

KoBeWi commented on editor/editor_settings.cpp:
```diff
@@ -68,6 +69,9 @@ bool EditorSettings::_set(const StringName &p_name, const Variant &p_value) {
 		changed_settings.insert(p_name);
 		emit_signal(SNAME("settings_changed"));
+		if (p_name == "text_editor/external/exec_path") {
+			_exec_path_changed(p_value);
```
This changes the path again, so it should happen before emitting the signal.

van800 replied:
It is not changing the path, it should change the `exec_flags`.
Where would you suggest to place it?
<img width="726" alt="image" src="https://github.com/user-attachments/assets/23b2079f-bf0c-4323-a71e-d6d55d57237e" />

KoBeWi replied:
Ah right. But it's still modifying a setting, so set it before the signal.

van800 replied:
done, thank you for the pointer.

KoBeWi commented on editor/editor_settings.cpp:
```diff
@@ -67,6 +68,13 @@ bool EditorSettings::_set(const StringName &p_name, const Variant &p_value) {
 	bool changed = _set_only(p_name, p_value);
 	if (changed && initialized) {
 		changed_settings.insert(p_name);
+		if (p_name == "text_editor/external/exec_path") {
```
```suggestion
		if (p_name == SNAME("text_editor/external/exec_path")) {
```

KoBeWi commented:
Yes.
There is one comment you missed, but it's not blocking.

van800 commented:
@Calinou Can we please have this merged?

Repiteo commented:
Thanks! Congratulations on your first merged contribution! 🎉

AR-DEV-1 commented on editor/editor_settings.cpp:
```diff
@@ -1108,6 +1116,40 @@ void EditorSettings::_load_default_visual_shader_editor_theme() {
+	regex->compile(editor_pattern);
+	Ref<RegExMatch> editor_match = regex->search(p_path.to_lower());
+
+	if (editor_match.is_null()) {
```
That is weird, I don't think that it was tested if you encountered this. Can you send your builds so I can test as well?

AR-DEV-1 replied:
cc @van800 @ivanskorikov
